### DISCUSSION TITLE
The easiest way to add support for mouse input.

### DISCUSSION THREAD

**Discussion by sannelo** [fnuecke/oc2#128] [Ideas]
The easiest way to add support for mouse input.

We just send mouse events (mouse coordinates, mouse key pressed).

(All this only happens when we use the keyboard block)

**Comment by sannelo** [comment]
You can do this as a "Redstone signal block", we send a request, and receive a response in the form of an array of pressed keys, mouse coordinates.

**Comment by lxhom** [comment]
Do you want to use [mouse tracking for terminals](https://www.xfree86.org/current/ctlseqs.html#Mouse%20Tracking) or a real input device like `/dev/input/mouse2`? The first one would probably make more sense if you're not running a whole X11 server on there. Providing your use case for mouse support might help a bit

**Reply by sannelo** [reply]
The first option, as in oc1.

**Reply by atirutw** [reply]
> The first option

This is very possible. In fact, some Linux programs(vim, for example) can use mouse input if the terminal supports it(GNOME Terminal, etc.).

**Reply by lxhom** [reply]
Yup, even though it might be a bit difficult because we have to map the area of the projector (assuming we're using a keyboard block like OP suggested) to pixels and send those to the terminal, that might be a lot easier in the computer block.

**Reply by sannelo** [reply]
With a projector, this is too difficult to do. You can simply duplicate images from the projector to the keyboard interface.

**Reply by AshtonSnapp** [reply]
can't wait to run i3wm with mouse support inside of Minecraft /s

**Reply by lxhom** [reply]
brb uploading my dotfiles into oc2

i wonder how hard itd be to actually do that, if you have a running x11 server you could just crosscompile and install picom (0.3mb) and i3-gaps (0.5mb), both are in C if im not completely wrong and a gcc toolchain for riscv exists, but the x11 server is probably going to be a tiny bit harder

the performance would be great tho

**Reply by spike11302000** [reply]
You will need a riscv musl toolchain which you can get [here](https://musl.cc/). and ya x11 might be difficult to get working, we need to use the x11 fbdev driver if im not mistaken other then getting x11 working just need to port what ever desktop environment you plan to use

**Reply by fnuecke** [reply]
buildroot can be configured to include X11 I believe, so feel free to mess around with that. Unlikely to make it into the default rootfs due to size and performance though.

**Comment by fnuecke** [comment]
If someone knows the magic escape sequences to send mouse input via terminal, let me know, might add that.

For the projector, since that uses a virtio keyboard, would probably also get  a virtio mouse dev, but reversing the actual pos without adding a UI would be... interesting (not impossible, but needs readback from GPU, and extra rendertargets with the uvs, so.. probably not worth it). And I'm not sure yet if I want to give the projector a UI. We'll see.

**Reply by bastetfurry** [reply]
I think if you simply implement the XTerm standard everyone is happy:
https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Normal-tracking-mode
